File: kos/csrc/pvr_dr.c
#include <dc/pvr.h>
#include <dc/sq.h>
#include <string.h>
#include <stdlib.h>
#include <malloc.h>

#define DR_MAX_VERTICES     65536
#define DR_MAX_HEADERS      1024

typedef struct {
    pvr_vertex_t    *vertices;
    pvr_poly_hdr_t  *headers;
    uint32_t        vertex_count;
    uint32_t        header_count;
    pvr_dr_state_t  dr_state;
    int             initialized;
} dr_state_t;

static dr_state_t dr = {0};

int __go_dr_init(void)
{
    if (dr.initialized)
        return 0;

    dr.vertices = (pvr_vertex_t *)memalign(32, DR_MAX_VERTICES * sizeof(pvr_vertex_t));
    if (!dr.vertices)
        return -1;

    dr.headers = (pvr_poly_hdr_t *)memalign(32, DR_MAX_HEADERS * sizeof(pvr_poly_hdr_t));
    if (!dr.headers) {
        free(dr.vertices);
        dr.vertices = NULL;
        return -1;
    }

    memset(dr.vertices, 0, DR_MAX_VERTICES * sizeof(pvr_vertex_t));
    memset(dr.headers, 0, DR_MAX_HEADERS * sizeof(pvr_poly_hdr_t));

    dr.initialized = 1;
    return 0;
}

void __go_dr_shutdown(void)
{
    if (dr.vertices) {
        free(dr.vertices);
        dr.vertices = NULL;
    }
    if (dr.headers) {
        free(dr.headers);
        dr.headers = NULL;
    }
    dr.initialized = 0;
}

void __go_dr_begin_frame(void)
{
    dr.vertex_count = 0;
    dr.header_count = 0;
    pvr_dr_init(&dr.dr_state);
}

void __go_dr_end_frame(void)
{
    pvr_dr_finish();
}

void *__go_dr_get_vertex(void)
{
    if (dr.vertex_count >= DR_MAX_VERTICES)
        return NULL;
    return &dr.vertices[dr.vertex_count++];
}

void *__go_dr_get_vertex_at(int index)
{
    if (index < 0 || index >= DR_MAX_VERTICES)
        return NULL;
    return &dr.vertices[index];
}

int __go_dr_get_vertex_count(void)
{
    return dr.vertex_count;
}

void __go_dr_set_vertex_count(int count)
{
    if (count >= 0 && count <= DR_MAX_VERTICES)
        dr.vertex_count = count;
}

void *__go_dr_get_header(void)
{
    if (dr.header_count >= DR_MAX_HEADERS)
        return NULL;
    return &dr.headers[dr.header_count++];
}

static inline void dr_submit_32(const void *data)
{
    pvr_vertex_t *dest = pvr_dr_target(dr.dr_state);
    memcpy(dest, data, 32);
    pvr_dr_commit(dest);
}

void __go_dr_submit_header(const void *hdr)
{
    dr_submit_32(hdr);
}

void __go_dr_submit_vertex(const void *vtx)
{
    dr_submit_32(vtx);
}

void __go_dr_submit_vertex_xyzc(uint32_t flags, float x, float y, float z, uint32_t argb)
{
    pvr_vertex_t *dest = pvr_dr_target(dr.dr_state);
    dest->flags = flags;
    dest->x = x;
    dest->y = y;
    dest->z = z;
    dest->u = 0.0f;
    dest->v = 0.0f;
    dest->argb = argb;
    dest->oargb = 0;
    pvr_dr_commit(dest);
}

void __go_dr_submit_vertex_full(uint32_t flags, float x, float y, float z,
                                 float u, float v, uint32_t argb, uint32_t oargb)
{
    pvr_vertex_t *dest = pvr_dr_target(dr.dr_state);
    dest->flags = flags;
    dest->x = x;
    dest->y = y;
    dest->z = z;
    dest->u = u;
    dest->v = v;
    dest->argb = argb;
    dest->oargb = oargb;
    pvr_dr_commit(dest);
}

void __go_dr_submit_vertices(int start, int end)
{
    if (start < 0 || end > (int)dr.vertex_count || start >= end)
        return;

    for (int i = start; i < end; i++) {
        dr_submit_32(&dr.vertices[i]);
    }
}

void __go_dr_submit_all_vertices(void)
{
    __go_dr_submit_vertices(0, dr.vertex_count);
}

void __go_dr_submit_strip(const void *hdr, const void *vertices, int vertex_count)
{
    const pvr_vertex_t *verts = (const pvr_vertex_t *)vertices;
    dr_submit_32(hdr);
    for (int i = 0; i < vertex_count; i++) {
        dr_submit_32(&verts[i]);
    }
}

int __go_dr_check_alignment(void)
{
    uintptr_t v = (uintptr_t)dr.vertices;
    uintptr_t h = (uintptr_t)dr.headers;
    if ((v & 31) != 0) return -1;
    if ((h & 31) != 0) return -2;
    return 0;
}

void *__go_dr_get_vertex_buffer(void)
{
    return dr.vertices;
}

void *__go_dr_get_header_buffer(void)
{
    return dr.headers;
}

File: kos/pvr_dr.go
//go:build dreamcast
// +build dreamcast

package kos

import "unsafe"

//extern __go_dr_init
func goDRInit() int32

//extern __go_dr_shutdown
func goDRShutdown()

//extern __go_dr_begin_frame
func goDRBeginFrame()

//extern __go_dr_end_frame
func goDREndFrame()

//extern __go_dr_get_vertex
func goDRGetVertex() unsafe.Pointer

//extern __go_dr_get_vertex_at
func goDRGetVertexAt(index int32) unsafe.Pointer

//extern __go_dr_get_vertex_count
func goDRGetVertexCount() int32

//extern __go_dr_set_vertex_count
func goDRSetVertexCount(count int32)

//extern __go_dr_get_header
func goDRGetHeader() unsafe.Pointer

//extern __go_dr_submit_header
func goDRSubmitHeader(hdr unsafe.Pointer)

//extern __go_dr_submit_vertex
func goDRSubmitVertex(vtx unsafe.Pointer)

//extern __go_dr_submit_vertices
func goDRSubmitVertices(start, end int32)

//extern __go_dr_submit_all_vertices
func goDRSubmitAllVertices()

//extern __go_dr_submit_vertex_xyzc
func goDRSubmitVertexXYZC(flags uint32, x, y, z float32, argb uint32)

//extern __go_dr_submit_vertex_full
func goDRSubmitVertexFull(flags uint32, x, y, z, u, v float32, argb, oargb uint32)

//extern __go_dr_submit_strip
func goDRSubmitStrip(hdr, vertices unsafe.Pointer, count int32)

//extern __go_dr_check_alignment
func goDRCheckAlignment() int32

//extern __go_dr_get_vertex_buffer
func goDRGetVertexBuffer() unsafe.Pointer

//extern __go_dr_get_header_buffer
func goDRGetHeaderBuffer() unsafe.Pointer

type DirectRenderer struct {
	initialized bool
}

var DR DirectRenderer

func (d *DirectRenderer) Init() error {
	if d.initialized {
		return nil
	}
	result := goDRInit()
	if result != 0 {
		return &DRError{Code: int(result), Message: "failed to allocate DR buffers"}
	}
	if goDRCheckAlignment() != 0 {
		return &DRError{Code: -3, Message: "DR buffers not properly aligned"}
	}
	d.initialized = true
	return nil
}

func (d *DirectRenderer) Shutdown() {
	if d.initialized {
		goDRShutdown()
		d.initialized = false
	}
}

func (d *DirectRenderer) BeginFrame() {
	goDRBeginFrame()
}

func (d *DirectRenderer) EndFrame() {
	goDREndFrame()
}

func (d *DirectRenderer) GetVertex() *PvrVertex {
	ptr := goDRGetVertex()
	if ptr == nil {
		return nil
	}
	return (*PvrVertex)(ptr)
}

func (d *DirectRenderer) GetVertexAt(index int) *PvrVertex {
	ptr := goDRGetVertexAt(int32(index))
	if ptr == nil {
		return nil
	}
	return (*PvrVertex)(ptr)
}

func (d *DirectRenderer) GetVertexCount() int {
	return int(goDRGetVertexCount())
}

func (d *DirectRenderer) SetVertexCount(count int) {
	goDRSetVertexCount(int32(count))
}

func (d *DirectRenderer) GetHeader() *PvrPolyHdr {
	ptr := goDRGetHeader()
	if ptr == nil {
		return nil
	}
	return (*PvrPolyHdr)(ptr)
}

func (d *DirectRenderer) SubmitHeader(hdr *PvrPolyHdr) {
	goDRSubmitHeader(unsafe.Pointer(hdr))
}

func (d *DirectRenderer) SubmitVertex(v *PvrVertex) {
	goDRSubmitVertex(unsafe.Pointer(v))
}

func (d *DirectRenderer) SubmitVertexXYZC(flags uint32, x, y, z float32, argb uint32) {
	goDRSubmitVertexXYZC(flags, x, y, z, argb)
}

func (d *DirectRenderer) SubmitVertexFull(flags uint32, x, y, z, u, v float32, argb, oargb uint32) {
	goDRSubmitVertexFull(flags, x, y, z, u, v, argb, oargb)
}

func (d *DirectRenderer) SubmitVertices(start, end int) {
	goDRSubmitVertices(int32(start), int32(end))
}

func (d *DirectRenderer) SubmitAllVertices() {
	goDRSubmitAllVertices()
}

func (d *DirectRenderer) SubmitStrip(hdr *PvrPolyHdr, vertices []PvrVertex) {
	if len(vertices) == 0 {
		return
	}
	goDRSubmitStrip(unsafe.Pointer(hdr), unsafe.Pointer(&vertices[0]), int32(len(vertices)))
}

func (d *DirectRenderer) Triangle(hdr *PvrPolyHdr,
	x1, y1, x2, y2, x3, y3 float32, z float32, color uint32) {
	goDRSubmitHeader(unsafe.Pointer(hdr))
	goDRSubmitVertexXYZC(PVR_CMD_VERTEX, x1, y1, z, color)
	goDRSubmitVertexXYZC(PVR_CMD_VERTEX, x2, y2, z, color)
	goDRSubmitVertexXYZC(PVR_CMD_VERTEX_EOL, x3, y3, z, color)
}

func (d *DirectRenderer) Quad(hdr *PvrPolyHdr,
	x1, y1, x2, y2, x3, y3, x4, y4 float32, z float32, color uint32) {
	goDRSubmitHeader(unsafe.Pointer(hdr))
	goDRSubmitVertexXYZC(PVR_CMD_VERTEX, x1, y1, z, color)
	goDRSubmitVertexXYZC(PVR_CMD_VERTEX, x2, y2, z, color)
	goDRSubmitVertexXYZC(PVR_CMD_VERTEX, x3, y3, z, color)
	goDRSubmitVertexXYZC(PVR_CMD_VERTEX_EOL, x4, y4, z, color)
}

func (d *DirectRenderer) CheckAlignment() error {
	result := goDRCheckAlignment()
	if result == 0 {
		return nil
	}
	return &DRError{Code: int(result), Message: "buffer alignment check failed"}
}

func (d *DirectRenderer) GetVertexBuffer() uintptr {
	return uintptr(goDRGetVertexBuffer())
}

func (d *DirectRenderer) GetHeaderBuffer() uintptr {
	return uintptr(goDRGetHeaderBuffer())
}

type DRError struct {
	Code    int
	Message string
}

func (e *DRError) Error() string {
	return e.Message
}
